src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -388,15 +461,35 @@ private bool ShouldInvalidateServer(Exception exception)
                         code = (ServerErrorCode)writeConcernError.GetValue("code", -1).ToInt32();
                         message = writeConcernError.GetValue("errmsg", null)?.AsString;
 
-                        if (IsNotMasterOrRecovering(code, message))
+                        if (IsStateChangeError(code, message))
                         {
-                            return true;
+                            return !IsStaleStateChangeError(commandException.Result, out invalidatingResponseTopologyVersion);
                         }
                     }
                 }
             }
 
+            invalidatingResponseTopologyVersion = null;
             return false;
+
+            bool IsStaleStateChangeError(BsonDocument response, out TopologyVersion? nonStaleResponseTopologyVersion)
+            {
+                if (_connectionPool.Generation > connection.Generation)
+                {
+                    // stale generation number
+                    nonStaleResponseTopologyVersion = null;
+                    return true;
+                }
+
+                var responseTopologyVersion = TopologyVersion.FromMongoCommandResponse(response);
+                bool isStale = description.TopologyVersion.IsFresherThanOrEqualToServerResponse(responseTopologyVersion);
+
+                nonStaleResponseTopologyVersion = isStale ? null : responseTopologyVersion;
+                return isStale;
+                // We use FresherThanOrEqualTo because a state change should come with a new topology version
+                // This is equivalent to description.TopologyVersion.CompareFreshnessToServerResponse(responseTopologyDescriptionVersion) >= 0;
+            }
+
```

Comment:
From the this part of the pseudo code in the [spec commit](https://github.com/mongodb/specifications/commit/7dd9c008be6edaf65368787d5e91c2ac3aeb95b8#diff-62b4da2e38e7d2a9c1a2d92d315895bbR1401-R1411)
```
  def isStaleError(topologyDescription, response):
        currentServer = topologyDescription.servers[server.address]
        currentGeneration = currentServer.pool.generation
        generation = get connection generation from response
        if generation < currentGeneration:
            # Stale generation number.
            return True

        currentTopologyVersion = currentServer.topologyVersion
        # True if the current error's topologyVersion is greater than the server's
        return compareTopologyVersion(currentTopologyVersion, response.get("topologyVersion")) >= 0
```